File: from-java-to-kotlin/src/main/java/pl/braintelligence/E03_InstanceOf.java
package pl.braintelligence;

import java.util.List;
import java.util.stream.Collectors;

public class E03_InstanceOf {
    public static void main(String[] args) {

        // Instance of - String
        String str = "abc";

        if (str instanceof String) {
            final String result = ((String) str).substring(1);
            System.out.println(result); // bc
        }

        // Instance of - List
        var list = List.of("1", "2", 3, 4);

        var result = list.stream()
                .filter(String.class::isInstance)
                .collect(Collectors.toList()); // 1, 2

        System.out.println(result);
    }
}

File: from-java-to-kotlin/src/main/java/pl/braintelligence/E03_InstanceOf.kt
package pl.braintelligence

fun main() {

    val str: String = "abc"

    // Instance of and smart cast
    if (str is String) {
        println(str.substring(1))
    }

    val list = listOf("1", "2", 3, 4)

    // One way - more Java way
    val result = list
            .filter { String::class.java.isInstance(it) }
            .run { print(this) } // bc

    // Another way - more Kotlin way
    list.filter { it is String }.run { print(this) } // bc

}
